plant - check history for min_brightness

ChristianKuehnel opened the pull request:
## Description:
You probably want to know if the place, where your plant is located at the moment is bright enough for the plant. Checking the min_brightness of a plant only makes sense when looking at the history. At night it's obviously dark, so it does not make sense to check min_brightness. Thus the new check looks at states in the database and checks if the maximum brightness over the last X days (where X can be configured) was above min_brightness. 


**Related issue (if applicable):** 


**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3417

## Example entry for `configuration.yaml` (if applicable):
```yaml
 plant:
   simulated_plant:
       sensors:
         moisture: sensor.mqtt_plant_moisture
         battery: sensor.mqtt_plant_battery
         temperature: sensor.mqtt_plant_temperature
         conductivity: sensor.mqtt_plant_conductivity
         brightness: sensor.mqtt_plant_brightness
       min_moisture: 20
       max_moisture: 60
       min_battery: 17
       check_days : 3
```
The new field is  " "check_days".

Feel free to use the simulator for testing this change: https://github.com/ChristianKuehnel/hass_simulated_plant

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

ChristianKuehnel committed:
Added option to create a group for the plant and all of it's sensors so that they appear together in the UI

ChristianKuehnel committed:
fixed warnings from the hound

ChristianKuehnel committed:
added check for min_brightness over several days

ChristianKuehnel committed:
fixed hound complaints

pvizeli commented on homeassistant/components/plant.py:
```diff
@@ -97,6 +104,16 @@ def async_setup(hass, config):
+            group_name = plant_config[CONF_GROUP_NAME]
+            members = ['{}.{}'.format(DOMAIN, plant_name)]  # own entity id
+            members.extend(sensor_entity_ids)
+            hass.states.async_set('group.{}'.format(group_name), 'off', {
```
Use the group service calls...

ChristianKuehnel replied:
Not sure what you mean with "service calls". I changed that part to the way it's done in ```components/demo.py```

Is that what you proposed?

ChristianKuehnel committed:
1) added missing dependency on recorder
2) using group.Group instead of hass.states.async_set as requested by @pvizeli

ChristianKuehnel committed:
fixed pylint error in docstring

pvizeli commented on homeassistant/components/plant.py:
```diff
@@ -97,8 +105,18 @@ def async_setup(hass, config):
+            group_name = plant_config[CONF_GROUP_NAME]
+            members = ['{}.{}'.format(DOMAIN, plant_name)]  # own entity id
+            members.extend(sensor_entity_ids)
+            groups.append(group.Group.async_create_group(hass, group_name,
```
Look into `helper/entity_component.py` to look how you call the group services:
```
hass.components.group.set_group(....)
```
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/group.py#L147

ChristianKuehnel replied:
ok, changed it

ChristianKuehnel committed:
changed the way the groups are created

ChristianKuehnel committed:
fixed requirements issue

pvizeli commented on homeassistant/components/plant.py:
```diff
@@ -97,8 +103,16 @@ def async_setup(hass, config):
+        if CONF_GROUP_NAME in plant_config and \
+                plant_config[CONF_GROUP_NAME] is not None:
+            group_name = plant_config[CONF_GROUP_NAME]
+            members = ['{}.{}'.format(DOMAIN, plant_name)]  # own entity id
```
Instead to generate IDs that you not can sure that the core will use the same IDs, you should only "create" a empty group without entities and give the group name into Entiy object. You can use the `async_added_to_hass` callback to join the entity in the group after the core as written the real entity id to the object.

ChristianKuehnel replied:
Fixed. 

This part is a bit frustrating for me: This was the 4th time, I've implemented the way groups are created. I hope this now meets of your expectations.

pvizeli commented on homeassistant/components/plant.py:
```diff
@@ -206,14 +231,65 @@ def _update_state(self):
+        (ie self._brightness_updated == True). The result of the last check
+        is cached in self._brightness_problem.
+        """
+        from homeassistant.components.recorder.models import States
```
Look into history sensor how you should do that right. Short: Read that only on startup and store it into https://docs.python.org/2/library/collections.html#collections.deque. Allow user to choise the size of this. Now track the sensore state change and write it into this queue.

ChristianKuehnel replied:
While I understand your proposal, I'm not really happy with it:
1. I'm re-implementing functionality that is already implemented in the database: Store history of a value, get last 3 days of measurements.
1. While it reduces the database read operations, you spend additional memory to store the data. It's the classical trade-off between CPU time and RAM. In this case I prefer the database.
1. I do not care how many samples there are, I care about the last N days. So the queue might get quite long, if you have a high sampling rate on your sensor. To avoid that I would have to implement some smart way of compressin/downsampling the data. And that's just overkill for such a simple feature.

So all in all, I think the current solution is still the best option.

ChristianKuehnel replied:
Took me quite some time, but now the data is compressed, so that we're storing only one entry per day. That's enough for this use case as we only want to know the maximum brightness of the course of a day.

And the data is read from the database only on startup, not during normal operation. And only if the recorder is available, so this is optional.

ChristianKuehnel committed:
Changed the way the groups are implemented. This is proposal number 4...

PeteBa commented:
@ChristianKuehnel, this may be a stupid question but could you use a plant state-card to achieve part 1 and avoid all of the group creation stuff? Given that the plant entity already has all of the sensor states stored in its own attributes, you could just expose that via a plant state-card. @pvizeli Is that an allowed design approach ?

Reason, I'm asking is that I've done exactly this for my 12 miflora's. I now "hide" the 60-odd sensors and just have the 12 plant entities showing in the UI but with all of the problem identification information visible. A state-card does this without needing any modification to your current code. I've attached code/screenshot so you get the jist.

Anyway, feel free to file in the "thanks but ..." bucket, as reading the PR history, I suspect the last thing you want is more options !

The part 2 stuff looks great and look forward to trying it out. Thanks for all the efforts on this.

![plant](https://user-images.githubusercontent.com/16738570/31144998-4fba364a-a87a-11e7-9fcf-bd9e32f1b3fe.jpg)
[state-card-plant.zip](https://github.com/home-assistant/home-assistant/files/1353658/state-card-plant.zip)

pvizeli commented on homeassistant/components/plant.py:
```diff
@@ -244,4 +319,21 @@ def state_attributes(self):
+            members = [self.entity_id]
+            members.extend(list(self._config[CONF_SENSORS].values()))
+            self.hass.components.group.set_group(self._group_name,
+                                                 entity_ids=members)
```
You can that do more simple with: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/group/__init__.py#L169

ChristianKuehnel replied:
@pvizeli 

1) From looking at the code you linked, I'm not sure how you want to simplify this. Please be more specific.
2) This is already the 4th version I've implemented to create these groups. This is quite frustrating for me to keep changing such a simple feature without benefit for the users. So please be more specific in your code reviews and describe more clearly what you would like to have.
3) Why are there so many different ways to create groups in HASS? And if only one of these ways is correct, then please provide some documentation so that other developers use the right ones, especially update  "demo.py".

ChristianKuehnel replied:
@pvizeli 
Since I did not hear from you: I just tried to use "add=members" instead of "entity_ids"=members but that does not provide the results I wanted to have: add seems to add the elements to the group in a random order, but I want the plant component to be the first element in the list.

So I'm still not sure what you want...

ChristianKuehnel commented:
@PeteBa 
I really like your solution! That looks really cool. Honestly I'm not sure how to proceed with the groups...

(in case anyone wants to use the state card : here are the instructions to install it: https://home-assistant.io/developers/frontend_creating_custom_ui/)

ChristianKuehnel commented:
@PeteBa 
After using it for a while, I have to admit that  the state cards look much nicer than the groups. The only downside that I found so far is that I cannot click e.g. on the moisture and then see the history of the value. But that's a minor issue.

Are you planning to merge the state card?

ChristianKuehnel committed:
Data read from recorder only on startup.

We now only store one data point per day. If a recorder is configured, this data is initialized from the database. If not the list is empty on startup.

PeteBa commented:
@ChristianKuehnel , thanks for the feedback! I've done a second iteration that I think looks better and brings in the battery level icons as well. I would love to get this merged in if you are supportive. I would want to do a small modification to plant.py (i.e. expose Icon and Units names) so I dont have to hard-code in the state-card. I assume this can be done, I just have to figure out how 8). I'll get a new PR going for this.

![image](https://user-images.githubusercontent.com/16738570/31574549-02bcb298-b0ca-11e7-9a48-75c133420aa1.png)

ChristianKuehnel committed:
added missing documentation

ChristianKuehnel committed:
fixed typo in comment

ChristianKuehnel commented:
@pvizeli 
The tests are now passing again, please merge or let me know what you want changed.

ChristianKuehnel commented:
@pvizeli I haven't heard from you in a while. How do you want to proceed here?

balloob commented:
I like the idea from @PeteBa a lot. It is the same thing that we do for weather: it too contains a lot of data points and it is shown in the UI as 1 card. I think that we should pursue that.

ChristianKuehnel commented:
Hey @balloob ,

just to be sure: So you want me to remove the group feature from this PR? And only leave the brightness history feature in?

balloob commented:
Do you think that we need the brightness feature or, when we add a new card for plants, we can just query the database for the history? (so it is also ok between restarts)

ChristianKuehnel commented:
I'm not sure if I fully understand your proposal.

Functionality:
The history feature collects the maximum brightness over the last n days for every day. This will help you to decide if the plant gets enough light and if you need to move them. I need the history as it does not make any sense to look at the brightness at night. So the min_brightness check was quite useless so far.

Solution:
Initially I implemented this by querying the database when I wanted to update the value. After the comments from @pvizeli I changed the behavior: The database is only read on startup, after that I'm keeping the history in memory (one value per day). So this should be quite efficient.

So all in all, I would say the history still makes sense.

My question is: Do we want to keep the grouping feature in or shall I remove it?

PeteBa commented:
@ChristianKuehnel , I have translated the `state-card` from above into the new `ha-card` format based on feedback from @balloob  above and submitted as work-in-progress here (https://github.com/home-assistant/home-assistant-polymer/pull/583). I am still showing the original brightness attribute as per the current Plant component but can change as needed.

![image](https://user-images.githubusercontent.com/16738570/32457971-b03c1d22-c322-11e7-98dc-8a8ea83b05a5.png)

Unless I've missed something, the new `ha-card` approach doesn't let you consolidate a group of plants into a single card like the custom `state-card` from the earlier post. Instead, you get one card per plant regardless of whether grouped or not.

So pros and cons here as some users may prefer the current group view (i.e. list of plant entities with just state shown), some may want the details always visible as per this `ha-card` and yet others may want something in between, such as, the default state-card with the above showing as the more-info when tapped.

I would very much appreciate your thoughts.

andrey-git commented:
So what's the end goal?
If each attribute (c % lx ... ) is not a separate sensor you don't get a separate history (unless a hack on the UI side is done to pull it from attributes)

ChristianKuehnel commented:
@andrey-git 
My goal was to have all the values that belong to one plant together in one element in the UI. So that you immediately get an overview of how the plant is doing. This can either be done with **groups** (this PR) or via the **UI** (https://github.com/home-assistant/home-assistant-polymer/pull/583).

My 2 cents:
* The **group solution** has the advantages that you still get the history for each sensor and the state badges disappear from the long list at the top of the UI. But they do not look so nice.
* The **UI solution** has the advantage that it looks nicer and you get the color-coding of the values that are out of range. But you do not get the history and you have to hide the sensors manually.

So both options have their advantages and disadvantages...

PeteBa commented:
@ChristianKuehnel , that is a great summary of the design choice here. Note that history works in the ***UI solution*** just haven't submitted as it requires changes to the plant component. Specifically, adding the sensor entity_ids to the plant state attributes so they are available on the polymer side.

A "con" for the `ha-card` is that it doesn't combine into a single card when grouping multiple plants together. You still end up with one card per plant entity. So no way to "hide" the detail from the user that just wants to see the headline ok/problem indicator.

The middle-ground might be to make this the "more-info" card that comes up when when you tap on a plant. The user would still need to manually hide all of the underlying sensors but they would have a consolidated view and access to history from `more-info`. If someone (like you and me !) wants the full detail on their view then the `custom-state-card` from above is an option.

c727 commented:
you should make the sensors optional
```
sensors:
  - entity: sensor.mqtt_plant_moisture
    type: moisture
```

ChristianKuehnel commented:
@c727 
This would be a breaking change as existing users would have to migrate their existing configuration.

It's no problem to export the entity ids of the sensors as additional attribute like. As soon as we have a decision on the path we want to take here I can easily add a new attribute to the component:
```sensors : {"moisture":"sensor.mqtt_plant_moisture", ... }````

PeteBa commented:
Hey @ChristianKuehnel , just wanted to check how you got on with the frontend build ? For info, I have pushed an update that has the "history" working if you click on one of the sensor icons. I used your suggestion of `attrib.sensors : {"moisture":"sensor.mqtt_plant_moisture", ... }` as a plant.py "patch" to test this. I am curious of your thoughts.

ChristianKuehnel committed:
removed group fature

ChristianKuehnel commented:
@PeteBa Just tried to set up the frontend development environment but I failed following the instructions: https://github.com/home-assistant/home-assistant/issues/10546

Anyway I just removed the group feature from this PR and added a new attribute "sensors" to it. This attribute will list all the sensors from which we're getting the data. This should allow you to display their history.

Please let me know if this is what you wanted to have.

ChristianKuehnel committed:
Merge remote-tracking branch 'upstream/master' into dev

ChristianKuehnel committed:
added group dependency since it's still needed

ChristianKuehnel committed:
fixed bug: now "None" is no longer added to the DailyHistory

PeteBa commented:
@ChristianKuehnel , just tried your latest commit and I have access to the sensors which is great. Given that all of these options can work, I guess the main question really is which do you think best meets the goal ? I am happy to refine the existing card or to change it to a more-info card if you think that is better?

If we do go forward with the card then we would need to agree where things like icons/uom should get mastered (i.e. the sensor, the plant component or the front-end) as there is some duplication at the moment.

PeteBa commented:
@ChristianKuehnel , sorry I think I was out-of-sync with your comments. If you have the front-end built via `script/build_frontend` and you can confirm that the three files ( src/components/ha-cards.html, src/cards/ha-card-chooser.html and src/cards/ha-plant-card.html) have all been updated as per the front-end PR then you should be seeing the plant-card. You may need to "clear cache" in Developer Tools (or with Chrome you can also go through a new Incognito tab which can also work). Let me know if it isn't coming up.

ChristianKuehnel committed:
now also outputting unit of measurement if defined by the sensors

ChristianKuehnel committed:
removed iconss

ChristianKuehnel commented:
@PeteBa thx for the hint, now I got your new plant card working.

I like the plant-card the way it is now: You get a nice compact overview of the current values in the UI. And you even getting the history by clicking on the icons. I'm not sure what to show in a "more info" card...

Concerning architecture: 
* I would say the unit of measurement the plan component should read it from the sensors and forward it to the UI. The sensors are defining the units and so this should be the source.
* The icons are "only" used for the visualisation and are not related to the sensors or the logic. So I would decide on these in the plant-card.

That being said, I just added 2 patches: one is adding the units as attributes, the other is removing the icon-related logic.

ChristianKuehnel committed:
fixed line length

balloob commented on homeassistant/components/plant.py:
```diff
@@ -30,7 +31,9 @@
 ATTR_PROBLEM = 'problem'
+ATTR_SENSORS = 'sensors'
 PROBLEM_NONE = 'none'
+ATTR_MAX_BRIGHTNESS_HISTORY = 'max. brightness in history'
```
attributes should not have spaces or periods.

ChristianKuehnel replied:
ACK

balloob commented:
This can be merged when the one comment has been addressed.

PeteBa commented:
@ChristianKuehnel , follow-on from above, does it make sense to publish both versions of brightness? It might be simpler, and cover off review comment, to just make `brightness` attribute the historic version or maybe rename `max_brightness` if you want it more explicit. I would then only display that rather than both in the front-end. (if people want the real-time version then they could refer to the underlying sensor history).

I saw that your `_load_history_from_db` function assumes that a brightness sensor has been configured but the schema has it as optional so this might cause an error if undefined. Worth checking.

My front-end PR was inadvertently merged before we had finalised things so I have raised [a new PR here](https://github.com/home-assistant/home-assistant-polymer/pull/633) to continue. It supports the mastering of uom/icons as you suggest above. But needs the decision on `brightness` attribute name to finalise. Let me know your thoughts or any feedback.

Looks like we have a shot at getting this into the next release so let me know if you think we can finalise before Saturday.

andrey-git commented:
Don't use `unit_of_measurement` as a dict. It is a standard attribute. If you want to pass a dict of units use another name - unit**s**_of_measurement for example

balloob commented on homeassistant/components/plant.py:
```diff
@@ -179,10 +190,16 @@ def state_changed(self, entity_id, _, new_state):
         else:
             raise _LOGGER.error("Unknown reading from sensor %s: %s",
                                 entity_id, value)
+        if new_state.attributes.get(CONF_UNIT_OF_MEASUREMENT, None) \
```
- Use `ATTR_UNIT_OF_MEASUREMENT`
- `None` is the default second argument when none is passed in. Since you're not using the return value, you can just test if the key exists with `if ATTR_UNIT_OF_MEASUREMENT in new_state.attributes:`

ChristianKuehnel replied:
ACK

ChristianKuehnel replied:
ACK

balloob commented on homeassistant/components/plant.py:
```diff
@@ -192,28 +209,67 @@ def _update_state(self):
+                    result.append(self._check_min(sensor_name, value, params))
+                result.append(self._check_max(sensor_name, value, params))
+
+        result = filter(lambda x: x is not None, result)
```
Don't use `filter`.

```python
result = [entry for entry in result if entry is not None]
```

ChristianKuehnel replied:
ACK.

In case anyone is interested why: https://stackoverflow.com/questions/3013449/list-filtering-list-comprehension-vs-lambda-filter

balloob commented on homeassistant/components/plant.py:
```diff
@@ -237,11 +293,59 @@ def state_attributes(self):
-            ATTR_ICON: self._icon,
             ATTR_PROBLEM: self._problems,
+            ATTR_SENSORS: self._readingmap,
+            CONF_UNIT_OF_MEASUREMENT: self._unit_of_measurement,
```
As Andrey set, this is not allowed. you cannot use CONF_UNIT_OF_MEASUREMENT or ATTR_UNIT_OF_MEASUREMENT. Name it something else.

ChristianKuehnel replied:
ACK

ChristianKuehnel committed:
Implemented changes requested in code reviews.

These changes affect the interface to the UI:
* renamed attribute for units of measurement to "unit_of_measurement_dict"
* renamed attribute for maximum in brightness history to "max_brightness"

ChristianKuehnel committed:
only loading the history if a brightness sensor was configured

ChristianKuehnel commented:
I just fixed **all** of the issues addressed in the comments of the last two days, **except**:
I did not remove the brightness value. I can't give you a real reason, it's just my gut feeling that it should stay there.

@PeteBa Thx for pointing out the issue with ```_load_history_from_db```. I really missed that case.

ChristianKuehnel committed:
fixed testcase

PeteBa commented:
@ChristianKuehnel , I'm trying to get max_brightness to trigger a " low problem" but having difficulty. Doc suggest that `if max_brightness < CONF_MIN_BRIGHTNESS then "low  problem"` but CONF_MIN_BRIGHTNESS isn't stored in READINGS array so isn't picked up in _check_min(). Chances are its just me missing something here.

I think the historical view really add value for brightness and wonder if it would be better than the real-time check for max as well. The thing I dont like about real-time checks for brightness and temperature is that they assume that the problem has gone away just because it is approaching night. I would want to know if either metric was outside limits during the CHECK_DAYS period as plant was under stress and probably still recuperating. I would be up for discussing but probably for a future PR ?

ChristianKuehnel committed:
fixed stupid bug in check of brightness history

Also added test for this bug

ChristianKuehnel commented:
@PeteBa Thank you for finding yet another bug :sunglasses: I also fixed this one...

Concerning the question on checking the history for other values as well:
I have an email notification set up to tell me, if one of my plants is currently unhappy, so I do not really care about the history of the values:
```yaml
- id: plantstatus
  trigger:
    platform: state
    entity_id: <entity ids of my plants>
    to: problem
  action:
    - service: notify.my_email
      data_template:
        message: > 
          the plant {{trigger.entity_id}} has a problem: 
          {{trigger.to_state.attributes.problem}}
          
          current values:
            Moisture: {{trigger.to_state.attributes.moisture}}
            Conductivity: {{trigger.to_state.attributes.conductivity}}
            Brightness: {{trigger.to_state.attributes.brightness}}
            Temperature: {{trigger.to_state.attributes.temperature}}
            Battery: {{trigger.to_state.attributes.battery}}
            Brightness: {{trigger.to_state.attributes.brightness}}
```
Would that also work for your use case?

The exception here is the minimum brightness, as there it does not really make any sense to check it against a minimal value, as it's supposed to be dark at night. But checking against a maximum value again makes sense: right now it's too bright and the plant should be moved. In that case I would get an email.

I also was thinking of a smarter way to model the requirements of the plants, e.g. if the moisture (or conductivity) was above A in the last B days, the plant should still be fine. Or if the integral over the light absorbed by the plant is above C its still fine. But that adds a lot of assumptions about the requirements of the plants. As I do not really have a good idea what plants really need, I'm not able to create such a model. (And that's also the reason, why I'm using these sensors in the first place :wink: )

ChristianKuehnel commented:
@balloob I'm fine with the current solution, from my point of view this can be merged.

ChristianKuehnel committed:
added missing docstring

MartinHjelmare commented:
@ChristianKuehnel seems the tests for the db failed.

NovapaX commented:
These kind of entities with a lot of separate relative data points could profit from the more-info-panel tabs approach we've discussed here.
Pointing back to that thread to gather some possible use-cases to explore and take into account.
https://github.com/home-assistant/home-assistant-polymer/issues/627

ChristianKuehnel committed:
Fixed sporadic failure in test case.

Sometimes the component was created before the data was stored in the history. This lead to an empty history being read.

ChristianKuehnel committed:
removed unused import statement in testcase

ChristianKuehnel commented:
@MartinHjelmare thx for the hint, I did not check for the test results as all tests passed in my TravisCI instance. 

It was a nasty sporadic failure of the test case, caused by the wrong "block_till_done" statement in the test case. I rand the test a several times now and the problem seems to be gone now.

ChristianKuehnel committed:
reverted change to test case

ChristianKuehnel committed:
Changed startup behavior of the component.

No failed tests after 20 local test runs.

ChristianKuehnel committed:
added missing docstring

ChristianKuehnel commented:
Hmpf. The tests are still failing sporadically. So I guess this is some sort of concurrency issue between my component, hass and the database. The tests always work on my machine.

Does anyone have some hints one how to debug this problem?

PeteBa commented:
@ChristianKuehnel , hey, for what it is worth I can confirm that the tests are also passing on my dev machine and in travis-ci on my github.

I wonder if there is a mix of things going on here, a few things to check:
- Travis is highlighting a problem with the sqlalchemy query syntax. Not an expert on this but can you try changing the filter parameter to comma separated rather than using "and" as you might be picking up the python keyword. Default behaviour for filter is sqlalchemy `and_`:

```
query = session.query(States).filter(States.entity_id == entity_id.lower(), States.last_updated > start_date)).order_by(States.last_updated.asc())
```
- If that doesn't fix thigs then the Recorder may not have recorded the three state changes and the query not returning anything.  Only thing I could suggest here comes from the recorder test suite (tests/components/recorder/test_purge.py) where a small pause, `sleep(0.1)`, is used to give the recorder thread time to complete. Worth a try.

- If you want to see if it is an intermittent problem then you can close/open the PR to trigger a travis build.

ChristianKuehnel committed:
fixed tests

* added hass.start() to Setup
* fixed call parameters in constructor
* added time.sleep

ChristianKuehnel commented:
Hey @balloob , can you give some directions on how to debug/fix this sporadically failing build?

ChristianKuehnel committed:
Merge remote-tracking branch 'upstream/dev' into dev

ChristianKuehnel commented:
Hey @balloob , @pvizeli 
can anyone of you point me in the right direction in debugging this randomly failing tests? There seems to be a concurrency issue when running the tests against the recorder. The failure only occures during tests on Travis, but never in production use or when testing on my machine...

ChristianKuehnel committed:
Merge remote-tracking branch 'upstream/dev' into dev

ChristianKuehnel commented:
Sorry for the confusion, I had messed up my branches...

ChristianKuehnel commented:
@MartinHjelmare do you have any idea what to do with the flaky test here? Any idea on how to debug the issue?

The easy way would be: just  @unittest.skip the test case...

MartinHjelmare reviewed (commented):
I had a look at the code, but couldn't find anything. I'm not familiar with the db though.

balloob commented on tests/components/test_plant.py:
```diff
@@ -34,30 +47,145 @@ def __init__(self, state=None):
+            self.hass.block_till_done()
+        # wait for the recorder to really store the data
+        self.hass.data[recorder.DATA_INSTANCE].block_till_done()
+        sleep(0.1)
```
Not allowed to use `sleep`.

ChristianKuehnel replied:
fixed

balloob commented:
Python 3.4 support will be dropped this year… so time will fix this ;)

ChristianKuehnel committed:
Merge remote-tracking branch 'upstream/dev' into dev

ChristianKuehnel committed:
removed sleep

ChristianKuehnel committed:
fixed typo in variable name

ChristianKuehnel committed:
disabled loading from database as it's not stable at the moment and nobody knows why :(

ChristianKuehnel commented on tests/components/test_plant.py:
```diff
@@ -12,19 +21,23 @@
     'brightness': 987,
 }
 
+BRIGHTNESS_ENTITIY = 'sensor.mqtt_plant_brightness'
```
fixed

ChristianKuehnel commented:
Since nobody could help be debug the issues with the tests for the database-related problems, I disabled the database parts entirely: Right now plant is only collecting the history during runtime. This will resolve the issues with the unstable test case.

Now plant does __not__ bootstrap the history from the database on startup any more. So after every restart the history is gone. This is not perfect, but at least we have some history available.

ChristianKuehnel committed:
fixed flake8

MartinHjelmare commented on tests/components/test_plant.py:
```diff
@@ -34,30 +46,149 @@ def __init__(self, state=None):
+        This test can should only be executed if the loading of the history
+        is enabled via plant.ENABLE_LOAD_HISTORY.
+        """
+        if plant.ENABLE_LOAD_HISTORY:
```
It's better to use pytest skipif:
https://docs.pytest.org/en/latest/skipping.html#id1

ChristianKuehnel replied:
good point, changed the code

ChristianKuehnel committed:
now using pytest.mark.skipif to skip test

PeteBa commented:
@ChristianKuehnel , great that you found a way through on this. I think it is a fair compromise to get something out now and revisit full history when python 3.4 is dropped. I will run a few regression tests on the front-end card to make sure everything is still good and then re-open the polymer pr so that it can also be merged. Thanks.
